repository/product.go

```diff
@@ -25,26 +29,69 @@ func NewProductRepo(db *gorm.DB) *productRepoImpl {
 }
 
 func (r *productRepoImpl) FindAll() ([]model.Product, error) {
+	// Preload() -> to join table when querying data which is have association with another table
+	// key -> field which is referencing to another table
+	var listProduct []model.Product
+	result := r.db.Preload("Category").Find(&listProduct)
+	if result.Error != nil{
+		return nil, result.Error
+	}
 	// TODO: select all product with category
-	return []model.Product{}, nil
+	return listProduct, nil
 }
 
 func (r *productRepoImpl) FindByID(id int) (*model.Product, error) {
 	// TODO: select product with category and user
-	return nil, nil
+	// Preload() -> to join table when querying data which is have association with another table
+	// key -> field which is referencing to another table
+	var product model.Product
+	result := r.db.Preload("Category").Preload("User").Where("id = ?", id).First(&product)
+	if errors.Is(result.Error, gorm.ErrRecordNotFound){
+		return nil, apperror.ErrProductNotFound
+
+	}
+	return &product, nil
 }
 
 func (r *productRepoImpl) Insert(product *model.Product) error {
-	// TODO: create product and check pgConn.PgError with code 23503 FK Violation
+	if err := r.db.Create(product).Error; err != nil{
+			// TODO: create product and check pgConn.PgError with code 23503 FK Violation
+			// Custom error with PgErr
+		var pgErr *pgconn.PgError
+		if errors.As(err, &pgErr){
+		if pgErr.Code == pgerrcode.ForeignKeyViolation{
+			return apperror.ErrInvalidUserIdOrCategoryId
+		}
+		return err
+	}
+	}
 	return nil
 }
 
 func (r *productRepoImpl) Update(id int, product *model.Product) error {
 	// TODO: Update only specific fields name, price, and category_id
+	 if err := r.db.Model(&model.Product{}).Where("id = ?", id).Updates(model.Product{
```

Comment:
as I mention during mentoring, if we have passed a valid database model on the finisher method, then we don't need to call method `.Model(&model.Product{})` or `.Table("")`. Please try remove the method `.Model` and pass the pointer of model.Product to `.Updates` method.